python/cudf_polars/cudf_polars/containers/datatype.py

```diff
@@ -7,105 +7,107 @@
+        return plc.DataType(TIME_UNIT_TO_PLC_DATETIME_TYPE[dtype.time_unit])
+
+    if isinstance(dtype, pl.Duration):
+        return plc.DataType(TIME_UNIT_TO_PLC_DURATION_TYPE[dtype.time_unit])
+
+    if isinstance(dtype, pl.Time):
         raise NotImplementedError("Time of day dtype not implemented")
-    elif isinstance(dtype, pl.Datetime):
-        if dtype.time_unit == "ms":
-            return plc.DataType(plc.TypeId.TIMESTAMP_MILLISECONDS)
-        elif dtype.time_unit == "us":
-            return plc.DataType(plc.TypeId.TIMESTAMP_MICROSECONDS)
-        elif dtype.time_unit == "ns":
-            return plc.DataType(plc.TypeId.TIMESTAMP_NANOSECONDS)
-        assert dtype.time_unit is not None  # pragma: no cover
-        assert_never(dtype.time_unit)
-    elif isinstance(dtype, pl.Duration):
-        if dtype.time_unit == "ms":
-            return plc.DataType(plc.TypeId.DURATION_MILLISECONDS)
-        elif dtype.time_unit == "us":
-            return plc.DataType(plc.TypeId.DURATION_MICROSECONDS)
-        elif dtype.time_unit == "ns":
-            return plc.DataType(plc.TypeId.DURATION_NANOSECONDS)
-        assert dtype.time_unit is not None  # pragma: no cover
-        assert_never(dtype.time_unit)
-    elif isinstance(dtype, pl.String):
-        return plc.DataType(plc.TypeId.STRING)
-    elif isinstance(dtype, pl.Null):
-        # TODO: Hopefully
-        return plc.DataType(plc.TypeId.EMPTY)
-    elif isinstance(dtype, pl.List):
-        # Recurse to catch unsupported inner types
+
+    if isinstance(dtype, pl.List):
         _ = _from_polars(dtype.inner)
         return plc.DataType(plc.TypeId.LIST)
-    elif isinstance(dtype, pl.Struct):
-        # Recurse to catch unsupported field types
+
+    if isinstance(dtype, pl.Struct):
         for field in dtype.fields:
             _ = _from_polars(field.dtype)
         return plc.DataType(plc.TypeId.STRUCT)
-    else:
-        raise NotImplementedError(f"{dtype=} conversion not supported")
+
+    raise NotImplementedError(f"{dtype=} conversion not supported")
+
+
+@cache
+def _to_polars(dtype: plc.DataType) -> pl.DataType:
+    type_id = dtype.id()
+
+    if type_id in PLC_TO_POLARS_TYPE:
+        return PLC_TO_POLARS_TYPE[type_id]()
+
+    if type_id in PLC_DATETIME_TYPE_TO_TIME_UNIT:
+        return pl.Datetime(PLC_DATETIME_TYPE_TO_TIME_UNIT[type_id])
+
+    if type_id in PLC_DURATION_TYPE_TO_TIME_UNIT:
+        return pl.Duration(PLC_DURATION_TYPE_TO_TIME_UNIT[type_id])
+
+    # TODO: STRUCT and LIST support
+    raise NotImplementedError(f"{dtype=} conversion not supported")
 
 
 class DataType:
     """A datatype, preserving polars metadata."""
 
-    polars: pl.datatypes.DataType
+    polars: pl.DataType | pl.Field
     plc: plc.DataType
 
-    def __init__(self, polars_dtype: pl.DataType) -> None:
-        self.polars = polars_dtype
-        self.plc = _from_polars(polars_dtype)
+    def __init__(self, dtype: pl.Field | pl.DataType | plc.DataType) -> None:
```

Comment:
When/why would we need to accept a `pl.Field`?

Also if we need to eventually support constructing from a `plc.DataType`, could we make a dedicated `from_pylibcudf(cls, dtype: plc.DataType` classmethod?

Reply:
This was added to make one of the tests pass (specifically `test_unhandled_dtype_conversion_raises`). It has one test case that has a `pl.Field`.

Reply:
> could we make a dedicated from_pylibcudf(cls, dtype: plc.DataType classmethod?

Yes

Reply:
Thanks, no we don't need to accept a `pl.Field`.